cloudapi/logs.go

```diff
@@ -176,3 +176,186 @@ func (c *Config) StreamLogsToLogger(
+
+// LogtailStreamDialer is a streaming connection dialer with a logtail service.
+type LogtailStreamDialer interface {
+	DialContext(ctx context.Context, url string, headers http.Header) (LogtailStreamer, error)
+}
+
+// dialAdapter adapts a gorilla websocket dialer as a LogtailStreamDialer.
+type dialAdapter struct {
+	dialer *websocket.Dialer
+}
+
+func (a dialAdapter) DialContext(ctx context.Context, url string, headers http.Header) (LogtailStreamer, error) {
+	// We don't need to close the http body or use it for anything until we want to actually log
+	// what the server returned as body when it errors out
+	conn, _, err := a.dialer.DialContext(ctx, url, headers) //nolint:bodyclose
+	return conn, err
+}
+
+// dialContext returns an established websocket connection with the logtail stream service.
+func dialContext(ctx context.Context, c dialConfig) (*wsconn, error) {
+	wsc := &wsconn{
+		m:          sync.Mutex{},
+		conn:       nil,
+		dialer:     dialAdapter{dialer: websocket.DefaultDialer},
+		dialConfig: c,
+		afterfn:    time.After,
+		onRedialing: func() {
+		},
+	}
+	if err := wsc.dial(ctx); err != nil {
+		return nil, err
+	}
+	if wsc.dialConfig.refreshAfter > 0 {
+		go wsc.keepAlive(ctx, wsc.dialConfig.refreshAfter)
+	}
+	return wsc, nil
+}
+
+type wsconn struct {
+	m          sync.Mutex
+	conn       LogtailStreamer
+	dialer     LogtailStreamDialer
+	dialConfig dialConfig
+
+	// afterfn abstracts a timer
+	afterfn func(time.Duration) <-chan time.Time
+
+	// onRedialing is a callback invoked
+	// when a connection's refresh attempt is invoked.
+	onRedialing func()
+}
+
+type dialConfig struct {
+	url           string
+	headers       http.Header
+	retryAttempts int64
+	retryInterval time.Duration
+	refreshAfter  time.Duration
+}
+
+// OnRedialing sets a function to invoke
+// when the a new connection refresh is on going.
+func (wsc *wsconn) OnRedialing(f func()) {
+	wsc.onRedialing = f
+}
+
+func (wsc *wsconn) dial(ctx context.Context) error {
+	attempts := uint(1)
+	if wsc.dialConfig.retryAttempts > 0 {
+		attempts = uint(wsc.dialConfig.retryAttempts)
+	}
+	var newc LogtailStreamer
+	err := retry(attempts, wsc.dialConfig.retryInterval, func() (err error) {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			newc, err = wsc.dialer.DialContext(ctx, wsc.dialConfig.url, wsc.dialConfig.headers)
+		}
+		return
+	})
+	if err != nil {
+		return err
+	}
+
+	// if any, close the previous conn
+	if wsc.conn != nil {
+		_ = wsc.Close()
+	}
+	wsc.m.Lock()
+	wsc.conn = newc
+	wsc.m.Unlock()
+	return nil
+}
+
+// keepAlive keeps alive the websocket connection invoking a new dial
+// every interval defined from refreshAfter.
+func (wsc *wsconn) keepAlive(ctx context.Context, refreshAfter time.Duration) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-wsc.afterfn(refreshAfter):
+			wsc.onRedialing()
+			// error skipped to let it fail on reading
+			// or retry a new dial on the next timer firing
+			_ = wsc.dial(ctx)
```

Comment:
I'm wondering if we should handle keep alive with a read deadline instead, see [this comment](https://github.com/gorilla/websocket/issues/649#issuecomment-718186107). It seems like the suggested approach and a good idea to set it regardless to avoid connections hanging, or having to depend on `context` for it.